### PR TITLE
feat: annotation relevant lsp package

### PR DESCRIPTION
**Pull request by vadson71**

### CODE REVIEW

File: packages/context/src/fetch.ts [hunk 1]
```diff
@@ -25,9 +18,26 @@ export default async function fetch(
   // the init opts to add the proxy agent
   if (proxy) {
     init = Object.assign({}, init, {
-      agent: HttpsProxyAgent(proxy),
+      agent: new HttpsProxyAgent(proxy),
```
Comment (petermuessig) [inline packages/context/src/fetch.ts:21]: There was a good reason to use the `HttpsProxyAgent` without `new`. Did you verify this? I think it was related to the `https-proxy-agent` exposing this class in a wrong way. Maybe this has been fixed in the meantime...

File: package.json [hunk 2]
```diff
@@ -40,7 +40,7 @@
     "@types/deep-equal-in-any-order": "1.0.1",
     "@types/fs-extra": "9.0.11",
     "@types/klaw-sync": "6.0.0",
-    "@types/lodash": "4.14.168",
+    "@types/lodash": "4.14.166",
```
Comment (petermuessig) [inline package.json:43]: we do you go back to older versions of the types?

File: packages/fe/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
+[![npm (scoped)](https://img.shields.io/npm/v/@ui5-language-assistant/context.svg)](https://www.npmjs.com/package/@ui5-language-assistant/fe)
+
+# @ui5-language-assistant/fe
+
+UI5 language assistant extension with LSP for annotation-relevant building blocks
+
+## Installation
```
Comment (petermuessig) [inline packages/fe/README.md:7]: Once the modules become private ones, this section isn't needed anymore.

File: packages/fe/package.json [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    "clean": "rimraf ./lib ./coverage ./nyc_output",
+    "compile": "yarn run clean && tsc -p .",
+    "compile:watch": "tsc -p . --watch",
+    "lint": "eslint . --ext .ts --max-warnings=0 --ignore-path=../../.gitignore",
+    "test": "mocha",
+    "coverage": "nyc mocha",
+    "format:fix": "prettier --write \"**/*.@(js|ts|json|md)\" --ignore-path=.gitignore"
+  },
+  "publishConfig": {
+    "access": "public"
```
Comment (petermuessig) [inline packages/fe/package.json:46]: will you make these modules `private` in the future?

File: packages/fe/src/services/completion/providers/context-path.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  context,
+  prefix,
+}: UI5AttributeValueCompletionOptions): AnnotationTargetInXMLAttributeValueCompletion[] {
+  const ui5Property = getUI5PropertyByXMLAttributeKey(
+    attribute,
+    context.ui5Model
+  );
+
+  if (
+    ui5Property?.library === "sap.fe.macros" &&
```
Comment (petermuessig) [inline packages/fe/src/services/completion/providers/context-path.ts:47]: You could use the types SAP_FE_MACROS?

File: packages/fe/src/services/completion/providers/meta-path.ts [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    | PropertyPathInXMLAttributeValueCompletion
+  )[] = [];
+  const ui5Property = getUI5PropertyByXMLAttributeKey(
+    attribute,
+    context.ui5Model
+  );
+
+  if (
+    !(
+      ui5Property?.library === "sap.fe.macros" &&
```
Comment (petermuessig) [inline packages/fe/src/services/completion/providers/meta-path.ts:58]: Usage of type SAP_FE_MACROS

**Review by petermuessig — COMMENTED** [review]
Just a rough scan of the code - basically I'm fine although I can't judge on the whole code functionality wise. I'm a bit unsure with the proxy change - see my comment their. From my POV the `fe` module is plugged in nicely into the `server.js` with just adding the `externalCompletionItems`.

**Comment by petermuessig** [comment]
BTW: @Klaus-Keller - as you are a co-author you need to also sign the CLA.

File: test-packages/framework/api.d.ts [hunk 7]
```diff
@@ -6,3 +6,5 @@ export {
   ReadFileResult,
   TestFramework,
 } from "./src/api";
+
+export const CURSOR_ANCHOR;
```
Comment (mr25mr) [inline test-packages/framework/api.d.ts:10]: maybe simply remove from here and added after line 7

like this consumer see it like
```typescript
(alias) const CURSOR_ANCHOR: any
```
where it should be

```typescript
(alias) const CURSOR_ANCHOR: "⇶"
```

File: test-packages/framework/src/framework.ts [hunk 8]
```diff
@@ -97,9 +102,89 @@ export class TestFramework implements TestFrameworkAPI {
...
     await writeFile(filePath, content);
   }
+
+  /**
+   * Updates file content
+   * @param filePath - path of file to update
+   * @param newText - new text to be written to the file
+   * @param options - describes how the file should be updated
+   *    - insertAfter: if provided the text is inserted after that fragment
+   *    - replaceText: if provided the new text replaces that fragment
```
Comment (mr25mr) [inline test-packages/framework/src/framework.ts:115]: please also add doc for

`insertBefore`
and `doUpdatesAfter`

File: test-packages/framework/src/framework.ts [hunk 9]
```diff
@@ -97,9 +102,89 @@ export class TestFramework implements TestFrameworkAPI {
...
+   * If options is omitted then new text is added to the end of file
+   */
+  public async updateFileContent(
```
Comment (mr25mr) [inline test-packages/framework/src/framework.ts:118]: The idea is to have a write function which returns void and read function which returns required data. With this, we change that idea. Maybe we should keep that idea to be consistent.

For offset, we have `setter` and `getter` which should holds  up to date data. Now offset is not keep up to date. Pease use ` this.offset = offset;` to assign new offset.

File: test-packages/framework/src/framework.ts [hunk 10]
```diff
@@ -97,9 +102,89 @@ export class TestFramework implements TestFrameworkAPI {
...
+        newContent =
+          existingContent.slice(0, index) +
+          newText +
+          existingContent.slice(offset);
+      }
+    } else {
+      newContent = `${existingContent}\n${newText}`;
+    }
+
+    const offset = newContent.indexOf(CURSOR_ANCHOR);
```
Comment (mr25mr) [inline test-packages/framework/src/framework.ts:179]: here we could use getOffset.

File: test-packages/framework/src/framework.ts [hunk 11]
```diff
@@ -1,19 +1,24 @@
...
-  deleteCopy,
+  deleteProject,
   createCopy,
   fileExitsSync,
   print,
   removeProjectContent,
 } from "./utils";
 
 import { TestFrameworkAPI, ProjectInfo, Config, ReadFileResult } from "./types";
+import { repeat } from "lodash";
```
Comment (mr25mr) [inline test-packages/framework/src/framework.ts:19]: please maintain dependency in package.json.

File: packages/fe/package.json [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    "api.d.ts",
+    "src"
+  ],
+  "main": "lib/src/api.js",
+  "repository": "https://github.com/sap/ui5-language-assistant/",
+  "license": "Apache-2.0",
+  "typings": "./api.d.ts",
+  "dependencies": {
+    "lodash": "4.17.21"
+  },
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/package.json:19]: Some of them I'll move to dependencies, but not all, From most of them I import types only

File: test-packages/framework/src/framework.ts (hunk 9 above)
Comment (vadson71) [reply test-packages/framework/src/framework.ts:118]: I don't like the idea to hold some offset in class property. Offset is returned to consumer right after file update. Why we we need to hold it in framework instance object?

File: test-packages/framework/src/framework.ts (hunk 10 above)
Comment (vadson71) [reply test-packages/framework/src/framework.ts:179]: Not sure, why we need it

File: test-packages/framework/src/framework.ts (hunk 11 above)
Comment (vadson71) [reply test-packages/framework/src/framework.ts:19]: What do you mean? lodash is already in dependencies

File: test-packages/framework/src/framework.ts (hunk 9 above)
Comment (vadson71) [reply test-packages/framework/src/framework.ts:118]: Actually I doubt we need a class for the framework as such. There is no internal state in the object. Just projectInfo which is provided by consumer anyway. Rest of the class are just functions, which don't use internal state

File: packages/fe/package.json (hunk 4 above)
Comment (vadson71) [reply packages/fe/package.json:46]: Probably not. I don't see reasons why we should do it

File: packages/fe/README.md (hunk 3 above)
Comment (vadson71) [reply packages/fe/README.md:7]: It is not private module

File: packages/fe/api.d.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import { UI5ValidatorsConfig } from "@ui5-language-assistant/xml-views-validation";
+import type { Context } from "@ui5-language-assistant/context";
+
+export {} from "./src/api";
+export type {} from "./src/types";
+export type {} from "@sap-ux/project-access";
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/api.d.ts:6]: will remove

File: packages/fe/api.d.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+export function getCompletionItems(opts: {
+  context: Context;
+  textDocumentPosition: TextDocumentPositionParams;
+  document: TextDocument;
+  documentSettings: Settings;
+  cst: CstNode;
+  tokenVector: IToken[];
+  ast: XMLDocument;
+}): CompletionItem[];
+export function isAnnotationIssue<T extends BaseUI5XMLViewIssue>(
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/api.d.ts:20]: sure, will add

File: packages/fe/src/services/completion/providers/context-path.ts [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+    if (!isAbsolutePath && completionSegmentIndex > 0) {
+      // relative paths are not supported
+      return [];
+    }
+    if (expectedAnnotations.length + expectedTypes.length === 0) {
+      return [];
+    }
+    // if (!isPropertyPath && expectedAnnotations.length === 0) {
+    //   return [];
+    // }
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/providers/context-path.ts:80]: done

File: packages/fe/src/services/completion/providers/context-path.ts [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+        metadata,
+        expectedAnnotations,
+        expectedTypes,
+        isPropertyPath
+      );
+      suggestions.push(
+        ...roots.map((root) => ({
+          element: root,
+          isLastSegment: !isNextSegmentPossible(root),
+          completionSegmentIndex,
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/providers/context-path.ts:112]: done

File: test-packages/framework/api.d.ts (hunk 7 above)
Comment (vadson71) [reply test-packages/framework/api.d.ts:10]: done

File: test-packages/framework/src/framework.ts (hunk 8 above)
Comment (vadson71) [reply test-packages/framework/src/framework.ts:115]: done

File: packages/fe/src/services/completion/providers/context-path.ts [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+          milestones
+        );
+        suggestions.push(
+          ...possibleTargets.map((t) => {
+            const entityType =
+              t._type === "NavigationProperty" ? t.targetType : t.entityType;
+            return {
+              element: t,
+              isLastSegment: !isNextSegmentPossible(entityType, milestones),
+              completionSegmentIndex,
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/providers/context-path.ts:147]: done

File: packages/fe/src/services/completion/providers/context-path.ts [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+      NavigationProperty: "N",
+    };
+
+    return suggestions.map((suggestion) => {
+      const text = `${completionSegmentIndex === 0 ? "/" : ""}${
+        ["EntityContainer", "EntitySet", "Singleton"].includes(
+          suggestion.element._type
+        ) && completionSegmentIndex < 2
+          ? suggestion.element.fullyQualifiedName
+          : suggestion.element.name
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/providers/context-path.ts:168]: All metadata elements have name defined. Only entity container has optional name. But container as root is always suggested with fully qualified name.

Will add function to get text

File: packages/fe/src/services/completion/providers/context-path.ts [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+      }`;
+      return {
+        type: "AnnotationTargetInXMLAttributeValue",
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/providers/context-path.ts:171]: No need to use constant here. TS already controls it. I'll change interface definitions. Then completion will work for type everywhere

File: packages/fe/src/services/completion/providers/context-path.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/providers/context-path.ts:47]: done

File: packages/fe/src/services/completion/providers/meta-path.ts [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/providers/meta-path.ts:58]: done

File: packages/fe/src/services/completion/utils.ts [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+import { DocumentCstNode } from "@xml-tools/parser";
+import { XMLToken, XMLDocument } from "@xml-tools/ast";
+
+import { GetXMLViewCompletionsOpts } from "@ui5-language-assistant/xml-views-completion";
+import { getSuggestions } from "@xml-tools/content-assist";
+import { Context } from "@ui5-language-assistant/context";
+import { TextDocument } from "vscode-languageserver-textdocument";
+import { attributeValueProviders } from "./providers";
+import { UI5XMLViewAnnotationCompletion } from "../../types";
+import { Settings } from "@ui5-language-assistant/settings";
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/utils.ts:22]: Only types are consumed, that's why all these modules are mentioned in devDependencies only

File: packages/fe/src/services/completion/utils.ts [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+
+// export function getAffectedRange(commitCharacters, completeString: CompleteString): Range {
+//   let startStringLength = completeString.position.character - completeString.startString.length;
+//   let remainingStringLength = completeString.position.character + completeString.remainingString.length;
+
+//   const commitCharacter = commitCharacters && commitCharacters.length ? commitCharacters[0] : '';
+
+//   // limit to the left with commitCharacter or '/'
+//   const fragmentArray = completeString.startString.split(commitCharacter || '/');
+//   if (fragmentArray.length > 1) {
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/completion/utils.ts]: done

File: packages/fe/src/services/diagnostics/validators/missing-entity-set.ts [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+    if (typeof contextPath === "string") {
+      return [];
+    } else {
+      if (!entitySet) {
+        return [
+          {
+            kind: "MissingEntitySet",
+            issueType: ANNOTATION_ISSUE_TYPE,
+            message:
+              "EntitySet for the current view is missing in application manifest. Attribute value completion and diagnostics are disabled",
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/services/diagnostics/validators/missing-entity-set.ts:49]: agree. may be in a separate BLI

File: packages/fe/src/types/completion.ts [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+// export enum CompletionItemKind {
+//   Term = 'Term',
+//   EntityContainer = 'EntityContainer',
+//   EntitySet = 'EntitySet',
+//   EntityType = 'EntityType',
+//   NavigationProperty = 'NavigationProperty',
+//   ComplexType = 'ComplexType',
+//   Property = 'Property',
+//   Unknown = 'Unknown'
+// }
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/types/completion.ts]: done

File: packages/fe/src/utils/metadata.ts [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+import { getNextPossiblePathTargets } from "./path";
+
+export type AllowedTargetType =
+  | "EntitySet"
+  | "EntityType"
+  | "Singleton"
+  | "NavigationProperty"
+  | "Property";
+
+// export const allowedTargets: AllowedTargetType[] = ['EntitySet', 'EntityType', 'Singleton'];
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/utils/metadata.ts]: done

File: packages/fe/src/utils/metadata.ts [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+  element: EntityType | EntitySet | Singleton | undefined,
+  property?: string,
+  navigationProperty?: string
+): AnnotationBase[] {
+  const matchedAnnotations: AnnotationBase[] = [];
+
+  if (!element) {
+    return [];
+  }
+  const type: EntityType | undefined = getEntityTypeForElement(element);
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/utils/metadata.ts:118]: true. fixed

File: packages/fe/src/utils/metadata.ts [hunk 27]
```diff
@@ -0,0 +1,171 @@
...
+      if (annotation.term === term.fullyQualifiedName) {
+        matchedAnnotations.push(annotation);
+      }
+    }
+  }
+  return matchedAnnotations;
+}
+
+export function getAnnotationAppliedOnElement(
+  metadata: ConvertedMetadata,
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/src/utils/metadata.ts:153]: Removed

File: packages/fe/test/services/completion/providers/context-path-spec.ts [hunk 28]
```diff
@@ -0,0 +1,567 @@
+import { expect } from "chai";
+import { stub } from "sinon";
+// import * as chai from 'chai';
+// import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';
+// chai.use(jestSnapshotPlugin());
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/completion/providers/context-path-spec.ts]: done

File: packages/fe/test/services/completion/providers/context-path-spec.ts [hunk 29]
```diff
@@ -0,0 +1,567 @@
...
+
+  const annotationSnippetCDS = `
+      annotate service.Travel with @(
+          UI.Chart #sample1 :{
+              ChartType: #Bar
+          },
+      );
+    `;
+  before(async function () {
+    const timeout = 5 * 60000 + 8000; // 5 min for initial npm install + 8 sec
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/completion/providers/context-path-spec.ts:55]: Not sure, why we need extra 10 seconds after 5 mins timeout

File: packages/fe/test/services/completion/providers/context-path-spec.ts [hunk 30]
```diff
@@ -0,0 +1,567 @@
...
+        "Passenger",
+        "Booking",
+        "BookedFlights",
+        "Countries",
+        "BookingStatus",
+        "BookingSupplement",
+        "Airline",
+        "Flight",
+        "Supplement",
+        "FlightConnection",
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/completion/providers/context-path-spec.ts:232]: They are already sorted as needed. Sorting property is not reflected in the snapshot. Previous test contains it

File: packages/fe/test/services/completion/providers/filter-bar-spec.ts [hunk 31]
```diff
@@ -0,0 +1,167 @@
...
+    },
+  };
+
+  const annotationSnippetCDS = `
+      annotate service.Travel with @(
+          UI.Chart #sample1 :{
+              ChartType: #Bar
+          },
+      );
+    `;
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/completion/providers/filter-bar-spec.ts:51]: yes, right. I don't inspect annotation content

File: packages/fe/test/services/completion/providers/meta-path-spec.ts [hunk 32]
```diff
@@ -0,0 +1,584 @@
...
+      "Main.view.xml"
+    );
+
+    await framework.updateFileContent(
+      annoFileSegmentsCDS,
+      annotationSnippetCDS
+    );
+  });
+
+  const getCompletionResult = async (
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/completion/providers/meta-path-spec.ts:91]: Many specific things are consumed in the function. It's not worth reusing it

File: packages/fe/test/services/completion/providers/meta-path-spec.ts [hunk 33]
```diff
@@ -0,0 +1,584 @@
...
+          delete newDetails["customViews"];
+          const newContext = {
+            ...c,
+            manifestDetails: newDetails,
+          };
+          return newContext;
+        }
+      );
+      expect(result.length).to.eq(0);
+    });
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/completion/providers/meta-path-spec.ts:287]: removed

File: packages/fe/test/services/completion/utils-spec.ts [hunk 34]
```diff
@@ -0,0 +1,9 @@
+import { expect } from "chai";
+import { getAffectedRange } from "../../../src/services/completion/utils";
+
+describe("completion utils", () => {
+  it("getAffectedRange edge use case", () => {
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/completion/utils-spec.ts:5]: It is implicitly tested in other tests

File: packages/fe/test/services/diagnostics/validators/missing-entity-set-spec.ts [hunk 35]
```diff
@@ -0,0 +1,166 @@
...
+    let result: AnnotationIssue[] = [];
+    try {
+      await framework.updateFileContent(viewFilePathSegments, snippet, {
+        insertAfter: "<content>",
+      });
+      const { cst, tokenVector } = await framework.readFile(
+        viewFilePathSegments
+      );
+      const context = await getContext(documentPath);
+      const xmlView = buildAst(cst as DocumentCstNode, tokenVector);
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/diagnostics/validators/missing-entity-set-spec.ts:73]: done

File: packages/fe/test/services/diagnostics/validators/missing-entity-set-spec.ts [hunk 36]
```diff
@@ -0,0 +1,166 @@
...
+          };
+          return newContext;
+        }
+      );
+      expect(result.map((item) => issueToSnapshot(item))).to.deep.equal([
+        "kind: MissingEntitySet; text: EntitySet for the current view is missing in application manifest. Attribute value completion and diagnostics are disabled; severity:info; offset:344-354",
+      ]);
+    });
+
+    it("service details are missing in manifest", async function () {
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/diagnostics/validators/missing-entity-set-spec.ts:111]: removed

File: packages/fe/test/services/diagnostics/validators/unknown-annotation-path-spec.ts [hunk 37]
```diff
@@ -0,0 +1,406 @@
...
+    );
+
+    await framework.updateFileContent(
+      annoFileSegmentsCDS,
+      annotationSnippetCDS
+    );
+    cache.reset(); // to refresh context
+  });
+
+  const validateView = async (
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/diagnostics/validators/unknown-annotation-path-spec.ts:78]: done

File: packages/fe/test/services/diagnostics/validators/unknown-annotation-path-spec.ts [hunk 38]
```diff
@@ -0,0 +1,406 @@
...
+
+    it("attribute value is absent", async function () {
+      const result = await validateView(
+        `<macros:Chart metaPath></macros:Chart>`,
+        this
+      );
+      expect(result.length).to.eq(0);
+    });
+
+    it("service details are missing in manifest", async function () {
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/diagnostics/validators/unknown-annotation-path-spec.ts:144]: removed

File: packages/fe/test/services/diagnostics/validators/unknown-annotation-path-spec.ts [hunk 39]
```diff
@@ -0,0 +1,406 @@
...
+            ...c,
+            manifestDetails: { ...c.manifestDetails, customViews: {} },
+          };
+          return newContext;
+        }
+      );
+      expect(result.length).to.eq(0);
+    });
+
+    it("wrong entitySet specified in manifestt", async function () {
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/diagnostics/validators/unknown-annotation-path-spec.ts:202]: fixed

File: packages/fe/test/services/diagnostics/validators/unknown-property-path-spec.ts [hunk 40]
```diff
@@ -0,0 +1,370 @@
...
+            ...c,
+            manifestDetails: newDetails,
+          };
+          return newContext;
+        }
+      );
+      expect(result.length).to.eq(0);
+    });
+
+    it("wrong entitySet specified in manifestt", async function () {
```
[the remark this answers is not in the document]
Comment (vadson71) [reply packages/fe/test/services/diagnostics/validators/unknown-property-path-spec.ts:237]: fixed. Yes some tests look similar, but different units are tested

File: test-packages/framework/src/framework.ts [hunk 41]
```diff
@@ -147,15 +255,43 @@ export class TestFramework implements TestFrameworkAPI {
...
+  }
+
+  /**
+   * Converts provided text into VSCode text document
+   * @param uri - file uri string
+   * @param content - file content
+   * @param offset - optional offset
+   * @returns - VSCode text document, document position parameters
+   */
+  public toVscodeTextDocument(
```
[the remark this answers is not in the document]
Comment (vadson71) [reply test-packages/framework/src/framework.ts:279]: done

File: test-packages/framework/src/framework.ts [hunk 42]
```diff
@@ -95,11 +117,98 @@ export class TestFramework implements TestFrameworkAPI {
...
+   *    - insertAfter: if provided, the new text is inserted after that fragment
+   *    - replaceText: if provided, the new text replaces that fragment
+   *    - doUpdatesAfter: if provided then search for that fragment is done first and
+   *                      if is found, then requested changes are made staring from position after it.
+   *                      If not found, exception is thrown
+   *
+   * If options is omitted then new text is added to the end of file
+   *
+   * returns object with offset of cursor anchor position
+   */
```
[the remark this answers is not in the document]
Comment (vadson71) [reply test-packages/framework/src/framework.ts:141]: done

**Review by mr25mr — APPROVED** [review]

File: packages/semantic-model/src/api-json.d.ts [hunk 43]
```diff
@@ -260,6 +260,12 @@ export interface Ui5Property {
   optional?: boolean;
   static?: boolean;
   allowedFor?: string[];
+  [ui5 - metadata]?: {
```
Comment (petermuessig) [inline packages/semantic-model/src/api-json.d.ts:263]: The UI5 team is also working on a semantic model for the `api.json` which could be reused in future...

**Review by petermuessig — COMMENTED** [review]
LGTM in general, only 1 thing which may be improved (`fetch.ts`) and a hint about the `api.json` schema.

@Klaus-Keller : you need to sign the CLA - otherwise this change can't be submitted!

**Comment by Klaus-Keller** [comment]
> BTW: @Klaus-Keller - as you are a co-author you need to also sign the CLA.

done

File: packages/semantic-model/src/api-json.d.ts (hunk 43 above)
Comment (vadson71) [reply packages/semantic-model/src/api-json.d.ts:263]: I guess this is just for information and nothing needs to be fixed for now

**Review by petermuessig — APPROVED** [review]
You can go ahead and merge if you want...

**Pull request merged**